### PR TITLE
Events

### PR DESCRIPTION
**Pull request by artyom-yurin**
### Description of the Change
Events

### Benefits
Part of Sealing

### Possible Drawbacks 
Tirggers ID

### CODE REVIEW

File: core/miner/storage_fsm/impl/events_impl.cpp [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+          auto handler = height_triggers_[tid];
+          if (handler.called) {
+            return outcome::success();
+          }
+
+          auto trigger_height = height - handler.confidence;
+
+          OUTCOME_TRY(income_tipset, tipset_cache_->getNonNull(trigger_height));
+
+          auto handle = handler.handler;
```
Comment (turuslan) [inline core/miner/storage_fsm/impl/events_impl.cpp:85]:
```suggestion
          auto &handle{handler.handler};
```

File: core/miner/storage_fsm/impl/events_impl.cpp [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+      auto maybe_error = tipset_cache_->add(change.value);
+      if (maybe_error.has_error()) {
+        logger_->error("Adding tipset into cache failed: {}",
+                       maybe_error.error().message());
+        return;
+      }
+
+      auto apply = [&](ChainEpoch height,
+                       const Tipset &tipset) -> outcome::result<void> {
```
Comment (turuslan) [inline core/miner/storage_fsm/impl/events_impl.cpp:74]:
```suggestion
      auto apply = [&](ChainEpoch height) -> outcome::result<void> {
```

File: core/miner/storage_fsm/impl/events_impl.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        for (const auto tid : height_to_trigger_[height]) {
+          auto handler = height_triggers_[tid];
```
Comment (turuslan) [inline core/miner/storage_fsm/impl/events_impl.cpp:76]:
```suggestion
          auto &handler{height_triggers_.at(tid)};
```

**Review by Alexey-N-Chernyshov — APPROVED** [review]

File: core/miner/storage_fsm/impl/events_impl.cpp (hunk 3 above)
Comment (artyom-yurin) [reply core/miner/storage_fsm/impl/events_impl.cpp]: Fixed

File: core/miner/storage_fsm/impl/events_impl.cpp (hunk 2 above)
Comment (artyom-yurin) [reply core/miner/storage_fsm/impl/events_impl.cpp]: Fixed

File: core/miner/storage_fsm/impl/events_impl.cpp (hunk 1 above)
Comment (artyom-yurin) [reply core/miner/storage_fsm/impl/events_impl.cpp]: Fixed

**Review by turuslan — APPROVED** [review]

**Pull request merged**